CMakeLists.txt

```diff
@@ -126,39 +134,56 @@ endif()
 # Update git submodules as needed
 #===============================================================================
 
+# Function that updates a git submodule, and double-checks that it exists
+function(update_git_submodule git_submodule_name)
+  execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init vendor/${git_submodule_name}
+                  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+                  RESULT_VARIABLE GIT_SUBMOD_RESULT)
+  if(NOT GIT_SUBMOD_RESULT EQUAL 0)
+    message(FATAL_ERROR "git submodule update --init failed with \
+      ${GIT_SUBMOD_RESULT} on ${git_submodule_name}, please checkout submodules")
+  endif()
+
+  # Double-check submodule indeed exists
+  if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/vendor/${git_submodule_name}/CMakeLists.txt")
+    message(FATAL_ERROR "The git submodule ${git_submodule_name} was not downloaded! GIT_SUBMODULE was \
+      turned off or failed. Please update the ${git_submodule_name} submodule and try again.")
+  endif()
+endfunction()
+
 find_package(Git)
 if(GIT_FOUND AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git")
   option(GIT_SUBMODULE "Check submodules during build" ON)
   if(GIT_SUBMODULE)
     message(STATUS "Submodule update")
-    execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive
-                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-                    RESULT_VARIABLE GIT_SUBMOD_RESULT)
-    if(NOT GIT_SUBMOD_RESULT EQUAL 0)
-      message(FATAL_ERROR "git submodule update --init failed with \
-        ${GIT_SUBMOD_RESULT}, please checkout submodules")
+    if (NOT fmt_FOUND)
+      update_git_submodule(fmt)
     endif()
+    if (NOT pugixml_FOUND)
+      update_git_submodule(pugixml)
+    endif()
+    update_git_submodule(gsl-lite)
+    update_git_submodule(xtensor)
+    update_git_submodule(xtl)
```

Comment:
Even if we do find fmt/pugixml, there's no harm in doing a submodule update for all submodules. I would personally just revert to what we had before in the interest of simplicity

Reply:
Hey Paul, I agree that this is maybe a violation of the YAGNI principle, but I was thinking we could maybe use this function for the optional libMesh dependency that @pshriwise is doing. I thought it would be useful for that--optionally downloading submodules. May be good as we add more bells and whistles going into the future...

Also, my reason for this approach solving this problem was that I'd like to ensure that if a system version is used, we 100% definitely are not including headers from a submodule directory. I found that causes failures with fmt, where I have v6 from OpenMC and v7 from my package manager which led to a compile failure.

Reply:
Longer term, we might want to use [`FetchContent`](https://cliutils.gitlab.io/modern-cmake/chapters/projects/fetch.html) instead of submodules anyway (especially for things like libmesh). Even if you do an optional submodule update here, that doesn't change the behavior if one does `git clone --recurse-submodules`. The submodule update doesn't cause OpenMC to install any headers from the dependencies (that happens with `add_subdirectory`, which you're already selectively avoiding).